las/credentials.py

```diff
@@ -1,68 +1,108 @@
 
-            if exists(credentials_path):
-                access_key_id, secret_access_key, api_key = self._read_from_file(credentials_path)
-            else:
-                access_key_id, secret_access_key, api_key = self._read_from_environ()
+        for idx, value in enumerate(creds):
+            if creds_from_environ[idx] is not None:
+                creds[idx] = creds_from_environ[idx]
 
-        if not all([access_key_id, secret_access_key, api_key]):
+        if not all(creds):
             raise MissingCredentials
 
-        self.access_key_id = access_key_id
-        self.secret_access_key = secret_access_key
+        client_id, client_secret, api_key, auth_endpoint, api_endpoint = creds
+
+        self.client_id = client_id
+        self.client_secret = client_secret
         self.api_key = api_key
+        self.auth_endpoint = auth_endpoint
+        self.api_endpoint = api_endpoint
 
     @staticmethod
-    def _read_from_environ() -> Tuple[Optional[str], Optional[str], Optional[str]]:
-        access_key_id = os.environ.get('LAS_ACCESS_KEY_ID')
-        secret_access_key = os.environ.get('LAS_SECRET_ACCESS_KEY')
-        api_key = os.environ.get('LAS_API_KEY')
-
-        return access_key_id, secret_access_key, api_key
+    def _read_from_environ() -> List[Optional[str]]:
+        return [os.environ.get(k) for k in (
+            'LAS_CLIENT_ID',
+            'LAS_CLIENT_SECRET',
+            'LAS_API_KEY',
+            'LAS_AUTH_ENDPOINT',
+            'LAS_API_ENDPOINT'
+        )]
 
     @staticmethod
-    def _read_from_file(credentials_path: str) -> Tuple[Optional[str], Optional[str], Optional[str]]:
+    def _read_from_file(credentials_path: str) -> List[Optional[str]]:
+        if not credentials_path:
+            credentials_path = expanduser('~/.lucidtech/credentials.cfg')
+
+        if not exists(credentials_path):
+            return [None] * 5
+
         config = configparser.ConfigParser()
         config.read(credentials_path)
         section = 'default'
 
-        access_key_id = config.get(section, 'access_key_id')
-        secret_access_key = config.get(section, 'secret_access_key')
+        client_id = config.get(section, 'client_id')
+        client_secret = config.get(section, 'client_secret')
         api_key = config.get(section, 'api_key')
+        auth_endpoint = config.get(section, 'auth_endpoint')
+        api_endpoint = config.get(section, 'api_endpoint')
+
+        return [client_id, client_secret, api_key, auth_endpoint, api_endpoint]
+
+    @property
+    def access_token(self) -> str:
+        access_token, expiration = self._token
+
+        if not access_token or time.time() > expiration:
+            access_token, expiration = self._get_client_credentials()
+            self._token = (access_token, expiration)
+
+        return access_token
+
+    def _get_client_credentials(self) -> Tuple[str, int]:
+        url = f'https://{self.auth_endpoint}/oauth2/token?grant_type=client_credentials'
```

Comment:
I suspect we'll need to change this line to be able to use mock endpoints as the 'oauth2' path part is specific to one of the endpoints we are using.
```python
url = f'https://{self.auth_endpoint}/oauth2/token?grant_type=client_credentials' # from this
url = f'https://{self.auth_endpoint}/token?grant_type=client_credentials' # to this
```

And update the endpoint to include 'oauth2' in the path in credentials file when needed.